### ISSUE TITLE
Crash on v.3.2.0: Fatal Exception: NSGenericException *** Collection was mutated while being enumerated.

### ISSUE THREAD

**Issue by madthijs** [AdColony/AdColony-iOS-SDK-3#33]
Hi,

Using AdColony SDK v3.2.0 we have the following crash
`[ADCLogManager setLogsActive] (ADCLogManager.m:93)`

Fabric reports:
```
Fatal Exception: NSGenericException
*** Collection <__NSArrayM: 0x18ab7980> was mutated while being enumerated.
```

Backtrace:
```
 # Crashlytics - plaintext stacktrace at Fri, 08 Sep 2017 08:29:54 GMT

#23. Crashed: com.twitter.crashlytics.ios.exception
0  APP                            0xa91ca7 CLSProcessRecordAllThreads + 11230375
1  APP                            0xa91ca7 CLSProcessRecordAllThreads + 11230375
2  APP                            0xa91b9f CLSProcessRecordAllThreads + 11230111
3  APP                            0xa85afd CLSHandler + 11180797
4  APP                            0xa90587 __CLSExceptionRecord_block_invoke + 11224455
5  libdispatch.dylib              0x1bf35783 _dispatch_client_callout + 22
6  libdispatch.dylib              0x1bf41a35 _dispatch_barrier_sync_f_invoke + 50
7  APP                            0xa8ff7b CLSExceptionRecord + 11222907
8  APP                            0xa8fda3 CLSExceptionRecordNSException + 11222435
9  APP                            0xa8f9a5 CLSTerminateHandler() + 11221413
10 libc++abi.dylib                0x1bae398f std::__terminate(void (*)()) + 78
11 libc++abi.dylib                0x1bae31a3 __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code, _Unwind_Exception*) + 110
12 libobjc.A.dylib                0x1baef139 _objc_exception_destructor(void*) + 246
13 CoreFoundation                 0x1c8675e1 -[NSException name] + 122
14 APP                            0xa08adf -[ADCLogManager setLogsActive] (ADCLogManager.m:93)
15 APP                            0x9f3c43 __32-[ADCNativeLayer loadController]_block_invoke_2 (ADCNativeLayer.m:142)
16 APP                            0x9c1dbd __62-[ADCController getInitialConfigAndDataWithCompletionHandler:]_block_invoke_2 (ADCController.m:317)
17 APP                            0x9c3cab __61-[ADCController downloadControllerWithURL:completionHandler:]_block_invoke (ADCController.m:599)
18 APP                            0x9d2c43 __75-[ADCDownloader initWithRequest:httpSession:destination:completionHandler:]_block_invoke (ADCDownloader.m:40)
19 CFNetwork                      0x1cd98aef __85-[__NSURLSessionLocal downloadTaskForRequest:downloadFilePath:resumeData:completion:]_block_invoke + 16
20 CFNetwork                      0x1cdf3c9b __51-[__NSCFLocalDownloadTask _private_errorCompletion]_block_invoke + 62
21 Foundation                     0x1d18db01 <redacted> + 8
22 Foundation                     0x1d0e3aef <redacted> + 146
23 Foundation                     0x1d0d5fe7 <redacted> + 770
24 Foundation                     0x1d18fd39 <redacted> + 190
25 libdispatch.dylib              0x1bf4259d _dispatch_queue_serial_drain + 854
26 libdispatch.dylib              0x1bf38b71 _dispatch_queue_invoke + 886
27 libdispatch.dylib              0x1bf441b5 _dispatch_root_queue_drain + 326
28 libdispatch.dylib              0x1bf4400f _dispatch_worker_thread3 + 106
29 libsystem_pthread.dylib        0x1c0ea87d _pthread_wqthread + 1040
30 libsystem_pthread.dylib        0x1c0ea45c start_wqthread + 8

--

Fatal Exception: NSGenericException
0  CoreFoundation                 0x1c867b3d __exceptionPreprocess
1  libobjc.A.dylib                0x1baef067 objc_exception_throw
2  CoreFoundation                 0x1c8675e1 -[NSException name]
3  APP                            0xa08adf -[ADCLogManager setLogsActive] (ADCLogManager.m:93)
4  APP                            0x9f3c43 __32-[ADCNativeLayer loadController]_block_invoke_2 (ADCNativeLayer.m:142)
5  APP                            0x9c1dbd __62-[ADCController getInitialConfigAndDataWithCompletionHandler:]_block_invoke_2 (ADCController.m:317)
6  APP                            0x9c3cab __61-[ADCController downloadControllerWithURL:completionHandler:]_block_invoke (ADCController.m:599)
7  APP                            0x9d2c43 __75-[ADCDownloader initWithRequest:httpSession:destination:completionHandler:]_block_invoke (ADCDownloader.m:40)
8  CFNetwork                      0x1cd98aef __85-[__NSURLSessionLocal downloadTaskForRequest:downloadFilePath:resumeData:completion:]_block_invoke
9  CFNetwork                      0x1cdf3c9b __51-[__NSCFLocalDownloadTask _private_errorCompletion]_block_invoke
10 Foundation                     0x1d18db01 (Missing)
11 Foundation                     0x1d0e3aef (Missing)
12 Foundation                     0x1d0d5fe7 (Missing)
13 Foundation                     0x1d18fd39 (Missing)
14 libdispatch.dylib              0x1bf4259d _dispatch_queue_serial_drain
15 libdispatch.dylib              0x1bf38b71 _dispatch_queue_invoke
16 libdispatch.dylib              0x1bf441b5 _dispatch_root_queue_drain
17 libdispatch.dylib              0x1bf4400f _dispatch_worker_thread3
18 libsystem_pthread.dylib        0x1c0ea87d _pthread_wqthread
19 libsystem_pthread.dylib        0x1c0ea45c start_wqthread

=========== REDACTED ============

#20. AdColony Worker
0  libsystem_kernel.dylib         0x1c020900 mach_msg_trap + 20
1  libsystem_kernel.dylib         0x1c0206e1 mach_msg + 44
2  CoreFoundation                 0x1c823be3 __CFRunLoopServiceMachPort + 144
3  CoreFoundation                 0x1c822065 __CFRunLoopRun + 1436
4  CoreFoundation                 0x1c7751af CFRunLoopRunSpecific + 470
5  CoreFoundation                 0x1c774fd1 CFRunLoopRunInMode + 104
6  Foundation                     0x1d0c9af5 <redacted> + 258
7  APP                            0xa2e4c5 -[AdColony setupWorkerThread] (AdColony.m:155)
8  Foundation                     0x1d1ad8eb <redacted> + 1122
9  libsystem_pthread.dylib        0x1c0ec93b _pthread_body + 216
10 libsystem_pthread.dylib        0x1c0ec85d _pthread_start + 234
11 libsystem_pthread.dylib        0x1c0ea468 thread_start + 8

```

Thanks and let me know if you need more info.
This is in production, I don't have a way to reproduce the crash (yet).

Thijs

**Comment by cmanibog-ac** [comment]
Our team has opened a bug for this crash and will be working on a resolution.

**Comment by anandjp** [comment]
Thanks for reporting it @madthijs. 
This issue has been fixed in 3.2.1 SDK.
